src/Microsoft.DotNet.ServiceFabric.ServiceHost/ServiceHostEventListener.cs

```diff
@@ -0,0 +1,147 @@
+using System;
+using System.Collections;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Collections.ObjectModel;
+using System.Diagnostics.Tracing;
+using System.Linq;
+using System.Text;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.DotNet.ServiceFabric.ServiceHost;
+
+public class ServiceHostEventListener : EventListener
+{
+    private static ServiceHostEventListener _instance;
+    private static IImmutableList<string> _eventSourceNamePrefixes;
+    private static ILoggerFactory _loggerFactory;
+    private static ConcurrentDictionary<string, ILogger> _loggers = new();
+
+    public static ServiceHostEventListener ListenToEventSources(ILoggerFactory loggerFactory, params string[] namePrefixes)
+    {
+        if (_instance is not null)
+        {
+            throw new InvalidOperationException("Cannot call ServiceHostEventListener.ListenToEventSources more than once.");
+        }
+        // These fields must be static because the "OnEventSourceCreated" method gets called inside the base constructor, so instance fields aren't initialized when it runs
+        _eventSourceNamePrefixes = namePrefixes.ToImmutableList();
+        _loggerFactory = loggerFactory;
+        return _instance = new ServiceHostEventListener();
+    }
+
+
+    protected override void OnEventSourceCreated(EventSource eventSource)
+    {
+        if (_eventSourceNamePrefixes.Any(prefix => eventSource.Name.StartsWith(prefix)))
+        {
+            EnableEvents(eventSource, EventLevel.Warning);
+            _loggers.TryAdd(eventSource.Name, _loggerFactory.CreateLogger(eventSource.Name));
+        }
+        base.OnEventSourceCreated(eventSource);
+    }
+
+    protected override void OnEventWritten(EventWrittenEventArgs eventData)
+    {
+        var level = eventData.Level switch
+        {
+            EventLevel.LogAlways => LogLevel.Critical,
+            EventLevel.Critical => LogLevel.Critical,
+            EventLevel.Error => LogLevel.Error,
+            EventLevel.Warning => LogLevel.Warning,
+            EventLevel.Informational => LogLevel.Information,
+            EventLevel.Verbose => LogLevel.Trace,
+            _ => LogLevel.None,
+        };
+
+        if (!_loggers.TryGetValue(eventData.EventSource.Name, out var logger))
```

Comment:
This could be first, since it's an early exit.  We log a lot, so the earlier we can have the "get out of here", the better, I think.

Reply:
Yep. I'll move it.